web-api/src/mind.rs

```diff
@@ -17,31 +17,116 @@
+    seq::{IteratorRandom, SliceRandom},
+    thread_rng,
+};
+use serde::{de, Deserialize, Deserializer};
 use xayn_ai_coi::nan_safe_f32_cmp_desc;
 
+use crate::models::DocumentId;
+
+#[derive(Debug, Deserialize)]
+struct ViewedDocument {
+    document_id: DocumentId,
+    was_clicked: bool,
+}
+
+impl ViewedDocument {
+    fn new(document_id: DocumentId, was_clicked: bool) -> Self {
+        Self {
+            document_id,
+            was_clicked,
+        }
+    }
+}
+
+fn deserialize_viewed_documents<'de, D>(deserializer: D) -> Result<Vec<ViewedDocument>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    let viewed_documents = s.split(' ');
+    viewed_documents
+        .map(|viewed_document| -> Result<_, D::Error> {
+            let mut parts = viewed_document.split('-');
+            let document_id = parts
+                .next()
+                .ok_or_else(|| de::Error::custom("missing document id"))?;
+            let document_id = DocumentId::new(document_id).map_err(de::Error::custom)?;
+            let was_clicked = parts
+                .next()
+                .ok_or_else(|| de::Error::custom("missing was_clicked"))?;
+            let was_clicked = match was_clicked {
+                "0" => Ok(false),
+                "1" => Ok(true),
+                _ => Err(de::Error::custom("invalid was_clicked")),
+            }?;
+            Ok(ViewedDocument::new(document_id, was_clicked))
+        })
+        .collect()
+}
+
+fn deserialize_clicked_documents<'de, D>(deserializer: D) -> Result<Vec<DocumentId>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    let viewed_documents = s.split(' ');
+    viewed_documents
+        .map(|document| -> Result<_, D::Error> {
+            let document_id = DocumentId::new(document).map_err(de::Error::custom)?;
+            Ok(document_id)
+        })
+        .collect()
+}
+
 #[derive(Debug, Deserialize)]
 struct Impression {
     id: String,
     user_id: String,
     time: String,
-    clicks: String,
-    news: String,
+    #[serde(deserialize_with = "deserialize_clicked_documents")]
+    clicks: Vec<DocumentId>,
+    #[serde(deserialize_with = "deserialize_viewed_documents")]
+    news: Vec<ViewedDocument>,
 }
 
-#[derive(Debug, Deserialize)]
-struct Article {
-    id: String,
+#[derive(Clone, Debug, Deserialize)]
+struct Document {
+    id: DocumentId,
     category: String,
     subcategory: String,
     title: String,
     snippet: String,
     url: String,
 }
 
-struct SnippetLabelPair(String, String);
+#[derive(Debug, Deserialize)]
+struct DocumentProvider {
+    articles: HashMap<DocumentId, Document>,
+}
+
+impl DocumentProvider {
+    fn new(path: &str) -> Result<Self, anyhow::Error> {
+        let articles = read::<Document>(path)?
+            .map(|article| article.map(|article| (article.id.clone(), article)))
+            .try_collect()?;
+        Ok(Self { articles })
+    }
```

Comment:
🔧 renaming should be done consistently (ff)
```suggestion
struct DocumentProvider {
    documents: HashMap<DocumentId, Document>,
}

impl DocumentProvider {
    fn new(path: &str) -> Result<Self, anyhow::Error> {
        let documents = read::<Document>(path)?
            .map(|document| document.map(|document| (document.id.clone(), document)))
            .try_collect()?;
        Ok(Self { documents })
    }
```